SignalServiceKit/src/Messages/Attachments/OWSAttachmentDownloads.m

```diff
@@ -80,99 +143,233 @@ - (instancetype)initWithAttachmentProtos:(NSArray<SSKProtoAttachmentPointer *> *
+                      @synchronized(attachmentStreams) {
+                          attachmentStreamsCopy = [attachmentStreams copy];
+                      }
+                      OWSLogInfo(@"Attachment downloads succeeded: %lu.", (unsigned long)attachmentStreamsCopy.count);
+
+                      dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                          successHandler(attachmentStreamsCopy);
+                      });
+                  })
+                  .catch(^(NSError *error) {
+                      OWSLogError(@"Attachment downloads failed.");
+
+                      dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+                          failureHandler(error);
+                      });
+                  });
+        [completionPromise retainUntilComplete];
+    });
+}
+
+- (void)enqueueJobForAttachmentPointer:(TSAttachmentPointer *)attachmentPointer
+                               message:(nullable TSMessage *)message
+                               success:(void (^)(TSAttachmentStream *attachmentStream))success
+                               failure:(void (^)(NSError *error))failure
+{
+    OWSAssertDebug(attachmentPointer);
+
+    OWSAttachmentDownloadJob *job = [[OWSAttachmentDownloadJob alloc] initWithAttachmentPointer:attachmentPointer
+                                                                                        message:message
+                                                                                        success:success
+                                                                                        failure:failure];
+
+    @synchronized(self) {
+        [self.attachmentDownloadJobQueue addObject:job];
     }
 
-    // We use PMKJoin(), not PMKWhen(), because we don't want the
-    // completion promise to execute until _all_ promises
-    // have either succeeded or failed. PMKWhen() executes as
-    // soon as any of its input promises fail.
-    AnyPromise *completionPromise
-        = PMKJoin(promises)
-              .then(^(id value) {
-                  NSArray<TSAttachmentStream *> *attachmentStreamsCopy;
-                  @synchronized(attachmentStreams) {
-                      attachmentStreamsCopy = [attachmentStreams copy];
-                  }
-                  OWSLogInfo(@"Attachment downloads succeeded: %lu.", (unsigned long)attachmentStreamsCopy.count);
-                  successHandler(attachmentStreamsCopy);
-              })
-              .catch(^(NSError *error) {
-                  failureHandler(error);
-              });
-    [completionPromise retainUntilComplete];
+    [self startDownloadIfPossible];
+}
+
+- (void)startDownloadIfPossible
+{
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+        OWSAttachmentDownloadJob *_Nullable job;
+
+        @synchronized(self) {
+            const NSUInteger kMaxSimultaneousDownloads = 4;
```

Comment:
WDYT?

Reply:
> WDYT 

Specific to this line, I think 4 concurrent downloads is a reasonable first guess.

More generally, the public API makes sense, and it's clear how it solves the problem. 

My instincts would have been to use NSOperation/NSOperationQueue/maxConcurrentOperations, rather than write our own synchronized throttling logic from scratch, but this should also work.